feat(bandwidth_scheduler) - include parent's receipts in bandwidth requests

jancionear opened the pull request:
When making a bandwidth request to a child shard which has been split from a parent shard, we have to include the receipts stored in the outgoing buffer to the parent shard in the bandwidth request for sending receipts to the child shard. Forwarding receipts from the buffer to parent uses bandwidth granted for sending receipts to one of the children. Not including the parent receipts in the bandwidth request could lead to a situation where a receipt can't be sent because the grant for sending receipts to a child is too small to send out a receipt from a buffer aimed at a parent.

jancionear commented on runtime/runtime/src/lib.rs:
```diff
@@ -2053,6 +2053,7 @@ impl Runtime {
         let pending_delayed_receipts = processing_state.delayed_receipts;
         let processed_delayed_receipts = process_receipts_result.processed_delayed_receipts;
         let promise_yield_result = process_receipts_result.promise_yield_result;
+        let shard_layout = epoch_info_provider.shard_layout(&apply_state.epoch_id)?;
```
This might break replayability, but AFAIU we now support only the last two versions.

wacban replied:
Why would it break replayability?

shreyan-gupta replied:
It's fine to have this here as long as we are not doing something different for past protocol versions.

wacban commented on runtime/runtime/src/congestion_control.rs:
```diff
@@ -564,7 +573,25 @@ impl ReceiptSinkV2 {
+            {
+                let parent_receipt_sizes_iter =
+                    parent_metadata.iter_receipt_group_sizes(trie, side_effects);
+                receipt_sizes_iter = Box::new(receipt_sizes_iter.chain(parent_receipt_sizes_iter));
```
I think resharding tries to empty the parent buffers first - (double check please) - so I would feel better to do it in the same order here. It may be important.

shreyan-gupta replied:
Agree, might start mattering what exactly do we fill within the bandwidth request as it's based on the order.

jancionear replied:
Ah good catch, I wanted to have the parent buffer first but I did the opposite by mistake

wacban reviewed (approved):
LGTM but see the comment about chaining order

wacban commented:
> Implementing it was the easy part, now how do I test it... I guess I could modify some resharding test and put a lot of big receipts that require bandwidth requests in the outgoing buffer to parent?

yeah, please see the testloop tests for resharding_v3, those should be usable for your case

shreyan-gupta reviewed (approved):
Looks good, thanks!

jancionear commented:
I added a `resharding_v3` test which sends 3MB receipts from one shard to a shard that is split into two children. The 3MB receipts are buffered in the buffer aimed at the parent shard, and bandwidth request generation needs to generate the proper requests to children in order to forward them. The are no receipts in the buffers to children to ensure that requests to children take into account the buffer to parent.

The situation looks like this:
```
new block #10 shards: [5, 3, 6] chunk mask [true, true, true] block hash QnHFtFZPfSGvQHxVB9G1JAhq4ZD1G4sibtA1kdAUy1D epoch id 2xnDZ56HJtq4ZkyYtqjcRoHyb6FdtMdVCAh46CUBFHhv
 1.129s  INFO test: outgoing buffers from shard 5: {}
 1.129s  INFO test: outgoing buffers from shard 3: {}
 1.129s  INFO test: outgoing buffers from shard 6: {}
 1.129s  INFO test: Sending 3MB receipt from account1 to account4. tx_hash: GqzwERTKRzRwAXfuSm1A4ywGqTxySRnNw6sEyn2MFNMD
 1.129s  INFO test: Sending 3MB receipt from account1 to account6. tx_hash: 6sPB9k96Rh7P8GsXT1jtsoLe7DnQbNuTJYa6GjJiqrWq
new block #11 shards: [5, 3, 6] chunk mask [true, true, true] block hash 37gf65nGFyumvuWL53uaViCgu7t9xHGKtjPDgf4bW3PH epoch id 2xnDZ56HJtq4ZkyYtqjcRoHyb6FdtMdVCAh46CUBFHhv
 1.203s  INFO test: outgoing buffers from shard 5: {}
 1.203s  INFO test: outgoing buffers from shard 3: {}
 1.203s  INFO test: outgoing buffers from shard 6: {}
 1.204s  INFO test: Sending 3MB receipt from account1 to account4. tx_hash: 6ULGwQoMeg6puNTqBK9pfqrPSKgk6Hjdag2bi3wWWLFz
 1.204s  INFO test: Sending 3MB receipt from account1 to account6. tx_hash: J5p1X9NGrUFUarbB3KNm3zjn2bqcJkUgRbhxWEYit85c
new block #12 shards: [5, 3, 6] chunk mask [true, true, true] block hash 4zYVCVQJjS8Dw9tPSRXwhXnkoU4DufmJAhUwhstSAi7r epoch id 2xnDZ56HJtq4ZkyYtqjcRoHyb6FdtMdVCAh46CUBFHhv
 1.380s  INFO test: outgoing buffers from shard 5: {}
 1.380s  INFO test: outgoing buffers from shard 3: {}
 1.380s  INFO test: outgoing buffers from shard 6: {}
 1.380s  INFO test: Sending 3MB receipt from account1 to account4. tx_hash: APiSRseffDPizHykMtbrU9NsNVuLeM7aMawU8x3gfYeT
 1.380s  INFO test: Sending 3MB receipt from account1 to account6. tx_hash: Hyvop8jvYfyUEui3dgfnGdQd6VwpXZX63pZos8Vh9PLz
new block #13 shards: [5, 3, 6] chunk mask [true, true, true] block hash AZPcFxXtF97cCL7gDEVG4QYFVUQmeoLWu3TYw2AYvig5 epoch id 2xnDZ56HJtq4ZkyYtqjcRoHyb6FdtMdVCAh46CUBFHhv
 1.932s  INFO test: outgoing buffers from shard 5: {}
 1.933s  INFO test: outgoing buffers from shard 3: {6: [3.0 MB, 3.0 MB]}
 1.933s  INFO test: outgoing buffers from shard 6: {}
new block #14 shards: [5, 3, 6] chunk mask [true, true, true] block hash 9qzQqmYPVgcck3T74HbyLjNPk8u4W1rx6BMzMZ1rU8e epoch id 2xnDZ56HJtq4ZkyYtqjcRoHyb6FdtMdVCAh46CUBFHhv
 2.759s  INFO test: outgoing buffers from shard 5: {}
 2.759s  INFO test: outgoing buffers from shard 3: {6: [3.0 MB]}
 2.759s  INFO test: outgoing buffers from shard 6: {}
new block #15 shards: [5, 3, 6] chunk mask [true, true, true] block hash yWrM9DLwivyW97thyBX8VAjVfpkTTEEYuLUQzNCdNZa epoch id 2xnDZ56HJtq4ZkyYtqjcRoHyb6FdtMdVCAh46CUBFHhv
 3.285s  INFO test: outgoing buffers from shard 5: {}
 3.287s  INFO test: outgoing buffers from shard 3: {6: [3.0 MB, 3.0 MB, 3.0 MB, 3.0 MB]}
 3.287s  INFO test: outgoing buffers from shard 7: {}
 3.287s  INFO test: outgoing buffers from shard 8: {}
new block #16 shards: [5, 3, 7, 8] chunk mask [true, true, true, true] block hash HSJ6nUBoSWRzaVkuLiT17TcfeXx7NJKKwzyyzS1VovNG epoch id 4N2CxMwS3G7uyYjXRPPNGgHzgxp1HZkJ3r8RZ1zYQAiU
 3.480s  INFO test: outgoing buffers from shard 5: {}
 3.482s  INFO test: outgoing buffers from shard 3: {6: [3.0 MB, 3.0 MB, 3.0 MB, 3.0 MB]}
 3.482s  INFO test: outgoing buffers from shard 7: {}
 3.482s  INFO test: outgoing buffers from shard 8: {}
new block #17 shards: [5, 3, 7, 8] chunk mask [true, true, true, true] block hash 7zr4maTR9BTzHU9uU6ciphy3UAxxN5WXpw58paHCapLQ epoch id 4N2CxMwS3G7uyYjXRPPNGgHzgxp1HZkJ3r8RZ1zYQAiU
 3.955s  INFO test: outgoing buffers from shard 5: {}
 3.957s  INFO test: outgoing buffers from shard 3: {6: [3.0 MB, 3.0 MB, 3.0 MB, 3.0 MB]}
 3.957s  INFO test: outgoing buffers from shard 7: {}
 3.957s  INFO test: outgoing buffers from shard 8: {}
new block #18 shards: [5, 3, 7, 8] chunk mask [true, true, true, true] block hash 9BQoNysb5fB8pf7cHfdnU5gfKjmwgnQodG3HZKVfC9Ae epoch id 4N2CxMwS3G7uyYjXRPPNGgHzgxp1HZkJ3r8RZ1zYQAiU
 4.731s  INFO test: outgoing buffers from shard 5: {}
 4.732s  INFO test: outgoing buffers from shard 3: {6: [3.0 MB, 3.0 MB, 3.0 MB]}
 4.732s  INFO test: outgoing buffers from shard 7: {}
 4.732s  INFO test: outgoing buffers from shard 8: {}
new block #19 shards: [5, 3, 7, 8] chunk mask [true, true, true, true] block hash A9knKwbPMujvHzd1qGjXBgPKvB6Et9orRT1ejdfGdWZy epoch id 4N2CxMwS3G7uyYjXRPPNGgHzgxp1HZkJ3r8RZ1zYQAiU
 5.515s  INFO test: outgoing buffers from shard 5: {}
 5.516s  INFO test: outgoing buffers from shard 3: {6: [3.0 MB, 3.0 MB]}
 5.516s  INFO test: outgoing buffers from shard 7: {}
 5.516s  INFO test: outgoing buffers from shard 8: {}
new block #20 shards: [5, 3, 7, 8] chunk mask [true, true, true, true] block hash AefEKam8WJ98qvLyrSczMXTGWwJuZQpsnwczWE9g68mX epoch id 4N2CxMwS3G7uyYjXRPPNGgHzgxp1HZkJ3r8RZ1zYQAiU
 6.293s  INFO test: outgoing buffers from shard 5: {}
 6.293s  INFO test: outgoing buffers from shard 3: {6: [3.0 MB]}
 6.293s  INFO test: outgoing buffers from shard 7: {}
 6.293s  INFO test: outgoing buffers from shard 8: {}
 new block #21 shards: [5, 3, 7, 8] chunk mask [true, true, true, true] block hash 2Zyb5eG975zvjSMeNofEZbYvLpSfucfAkJfy7Ar4DPHB epoch id 4N2CxMwS3G7uyYjXRPPNGgHzgxp1HZkJ3r8RZ1zYQAiU
 6.669s  INFO test: outgoing buffers from shard 5: {}
 6.669s  INFO test: outgoing buffers from shard 3: {}
 6.669s  INFO test: outgoing buffers from shard 7: {}
 6.669s  INFO test: outgoing buffers from shard 8: {}
```

The test actually found some bugs in the initial implementation. The main bug was that bandwidth requests were only generated for shards which had an outgoing buffer, meaning that there had to be at least one receipt aimed at the shard in the past. New shards created after resharding didn't have an outgoing buffer, and because of that requests were not generated for them. The other bug was that the code short-circuited when there were no receipts in the outgoing buffer. It made sense before, but now we also have to check the receipts in the parent buffer before we can decide that there's no need to make a request.

I rewrote `generate_bandwidth_request` to better reflect the new logic.

wacban reviewed (approved):
LGTM, thanks!

jancionear commented on integration-tests/src/test_loop/tests/resharding_v3.rs:
```diff
@@ -1028,6 +1028,47 @@ fn test_resharding_v3_buffered_receipts_towards_splitted_shard_v2() {
     test_resharding_v3_buffered_receipts_towards_splitted_shard_base(2);
 }
 
+/// This test sends large (3MB) receipts from a stable shard to shard that will be split into two.
+/// These large receipts are buffered and at the resharding boundary the stable shard's outgoing
+/// buffer contains receipts to the shard that was split. Bandwidth requests to the child where the
+/// receipts will be sent must include the receipts stored in outgoing buffer to the parent shard,
+/// otherwise there will be no bandwidth grants to send them.
```
Yes that's correct :+1:

jancionear commented on integration-tests/src/test_loop/utils/resharding.rs:
```diff
@@ -330,6 +334,143 @@ pub(crate) fn call_burn_gas_contract(
+                    let cur_epoch_estimated_end = cur_epoch_start + cur_epoch_length;
+                    cur_epoch_estimated_end
+                }
+                _ => tip.height + 99999999999999, // Not in the next epoch, set to infinity into the future
```
I initially did that, but then I had to add something to `estimated_resharding_height` and that'd overflow, so I changed it to something smaller, but still very far into the future. I think there is no more adding now, but I'd prefer to keep it the way it is, less likely to cause trouble in the future.

jancionear commented on integration-tests/src/test_loop/utils/resharding.rs:
```diff
@@ -330,6 +334,143 @@ pub(crate) fn call_burn_gas_contract(
+            for shard_uid in epoch_manager.get_shard_layout(&tip.epoch_id).unwrap().shard_uids() {
+                let mut outgoing_receipt_sizes: BTreeMap<ShardId, Vec<ByteSize>> = BTreeMap::new();
+
+                let memtrie =
+                    get_memtrie_for_shard(&client_actor.client, &shard_uid, &tip.prev_block_hash);
```
The argument is called `prev_block_hash`:

```rust
pub fn get_memtrie_for_shard(
    client: &Client,
    shard_uid: &ShardUId,
    prev_block_hash: &CryptoHash,
) -> Trie {
    let state_root =
        *client.chain.get_chunk_extra(prev_block_hash, shard_uid).unwrap().state_root();

    // Here memtries will be used as long as client has memtries enabled.
    let memtrie = client
        .runtime_adapter
        .get_trie_for_shard(shard_uid.shard_id(), prev_block_hash, state_root, false)
        .unwrap();
    assert!(memtrie.has_memtries());
    memtrie
}
```

AFAIU we get post-state root of the chunk included in the previous block, which means that it's the pre-state root of the chunk at the current height.
But tbh I just copied it from `check_buffered_receipts_exist_in_memtrie`, in this test being off by one doesn't really matter, what matters is that there are only receipts in the outgoing buffer to the parent shard.

jancionear commented on integration-tests/src/test_loop/utils/resharding.rs:
```diff
@@ -330,6 +334,143 @@ pub(crate) fn call_burn_gas_contract(
+                for target_shard in outgoing_buffers.shards() {
+                    let mut receipt_sizes = Vec::new();
+                    for receipt in outgoing_buffers.to_shard(target_shard).iter(&memtrie, false) {
+                        let receipt_size = match receipt {
+                            Ok(ReceiptOrStateStoredReceipt::StateStoredReceipt(
+                                state_stored_receipt,
+                            )) => state_stored_receipt.metadata().congestion_size,
+                            _ => panic!("receipt is {:?}", receipt),
+                        };
```
I feel like `let else` works well with short patterns, but for longer ones I find it harder to read than a standard `match`.

jancionear commented on integration-tests/src/test_loop/utils/resharding.rs:
```diff
@@ -330,6 +334,143 @@ pub(crate) fn call_burn_gas_contract(
+                    let cur_epoch_length =
+                        epoch_manager.get_epoch_config(&tip.epoch_id).unwrap().epoch_length;
+                    let cur_epoch_estimated_end = cur_epoch_start + cur_epoch_length;
+                    cur_epoch_estimated_end
```
Hmm you're right, `resharding_height` is the last block of old epoch, but `cur_epoch_start + cur_epoch_length` is actually the first block of the new one. I'll change it to match `resharding_height`.

jancionear commented on integration-tests/src/test_loop/utils/resharding.rs:
```diff
@@ -330,6 +334,143 @@ pub(crate) fn call_burn_gas_contract(
+            {
+                for signer_id in &signer_ids {
+                    for receiver_id in &receiver_ids {
+                        // Send a 3MB cross-shard receipt from signer_id's shard to receiver_id's shard.
```
I'd have to be a function with 8 arguments :/
`call_burn_gas_contract` also has the tx sending logic inlined, I think it's good enough.
